src/libraries/System.Private.CoreLib/src/System/IO/Strategies/BufferedFileStreamStrategy.cs

```diff
@@ -1067,7 +1069,8 @@ private void EnsureBufferAllocated()
 
             void AllocateBuffer() // logic kept in a separate method to get EnsureBufferAllocated() inlined
             {
-                _strategy.OnBufferAllocated(_buffer = new byte[_bufferSize]);
+                _buffer = GC.AllocateUninitializedArray<byte>(_bufferSize,
+                    pinned: true); // this allows us to avoid pinning when the buffer is used for the syscalls
```

Comment:
That correct, but `pinned: true` also allocates the array in Gen2 as side-effect so this may actually hurt real-world scenarios at the end..

Reply:
We can allocate it and use a GCHandle if that ends up being better. Previously it was pinned as part of a PreallocatedOverlapped.

(It's still not at all obvious when this newfangled POH should be used. )

Reply:
We can also investigate not pinning at all, in which case the code this interacts with will just create a gchandle for each operation.

And/or look at using a pool array, but we'd want to ensure enough synchronization was in place to minimize erroneous usage causing us to return an array still in use.  We do that in a few other streams.

Reply:
Here is a simple test:
```
using System;
using System.IO;

for (int i = 0; i < 100_000; i++)
{
    using (var f = new FileStream("test", FileMode.Create))
    {
        f.WriteByte(42);
    }
}
Console.WriteLine($"Allocated: {GC.GetTotalAllocatedBytes()} Gen2 GCs: {GC.CollectionCount(2)}");
```

- .NET 5: Allocated: 442474096 Gen2 GCs: 0
- This PR: Allocated: 448051624 Gen2 GCs: 103

It will be interesting to see whether these excessive Gen2 GCs hit performance gates of services trying .NET 6 previews.

> It's still not at all obvious when this newfangled POH should be used.

Agree. It is very hard to use.

> we'd want to ensure enough synchronization was in place to minimize erroneous usage causing us to return an array still in use

If you can cover all these cases, it may be better to use unmanaged buffer. It is pinned too, and it does not cause excessive Gen2 GCs.

Reply:
So allocating on the POH contributes to the gen2 budget. This is why we disable the buffer using size 1, that still works right?

Reply:
Nothing in this file is used at all if buffer size is 1.

Reply:
I'm looking forward to taking another stab at optimizing static files in .NET 6

Reply:
> If you can cover all these cases, it may be better to use unmanaged buffer. It is pinned too, and it does not cause excessive Gen2 GCs.

I'm going to start with a GCHandle and a normally allocated array.  I believe in that case I can mostly restrict synchronization to the async code paths (plus disposal).  If we use a native buffer, we'll need to protect the sync code paths as well.  We can start with this and then see if it makes sense to use a pooled or native buffer as well.

Reply:
Well, actually, I'm going to start by not pinning here at all (it'll then pin/unpin in the rest of the implementation per operation).  If there's no measurable impact, we can stick with that for now.